Clarify resourceVersion documentation

bprashanth opened the pull request:
Distil the conversation in https://github.com/GoogleCloudPlatform/kubernetes/issues/3086 into api-conventions.md, and include a link to the same in the description of resourceVersion.

bprashanth committed:
Clarify documentation around resourceVersion

bgrant0607 commented on docs/api-conventions.md:
```diff
@@ -190,9 +190,14 @@ achieve the state, and for the user to know what to anticipate.
 
-Read-modify-write consistency is accomplished with optimistic currency.
+Kubernetes leverages the concept of resource versions to achieve optimistic concurrency. When a record is about to be updated, it is checked against a pre-saved value (the version), and if it doesn't match, the update fails. In general, there are several strategies for versioning:
+* A numeric counter
```
I don't think describing strategies we aren't using is very helpful.

bgrant0607 commented on docs/api-conventions.md:
```diff
@@ -190,9 +190,14 @@ achieve the state, and for the user to know what to anticipate.
+* A database specific row id
 
-All resources have "resourceVersion" as part of their metadata. resourceVersion is a string that identifies the internal version of an object that can be used by clients to determine when objects have changed. It is changed by the server every time an object is modified. If resourceVersion is included with the PUT operation the system will verify that there have not been other successful mutation ...
+It's important to note that the application should *not* rely on the implementation details of the versioning system maintained by kubernetes, but on the patterns to achieve optimistic concurrency discussed below.
```
Capitalize Kubernetes.

bprashanth commented:
It was mentioned that calling the ObjectReference.ResourceVersion read-only is misleading so I've updated the PR by removing 'read-only' from its description. Also removed some newlines from the md. PTAL.

bgrant0607 commented:
Who mentioned that read-only was confusing? The client isn't allowed to change resourceVersion.

bgrant0607 commented on pkg/api/v1beta2/types.go:
```diff
@@ -861,7 +861,7 @@ type ObjectReference struct {
 	ID              string    `json:"name,omitempty" description:"id of the referent"`
 	UID             types.UID `json:"uid,omitempty" description:"uid of the referent"`
 	APIVersion      string    `json:"apiVersion,omitempty" description:"API version of the referent"`
-	ResourceVersion string    `json:"resourceVersion,omitempty" description:"specific resourceVersion to which this reference is made, if any"`
```
Please rebase.

bprashanth replied:
Done

bgrant0607 commented on docs/api-conventions.md:
```diff
@@ -190,15 +190,15 @@ achieve the state, and for the user to know what to anticipate.
 -----------------------------------
 
-Read-modify-write consistency is accomplished with optimistic currency.
+Kubernetes leverages the concept of *resource versions* to achieve optimistic concurrency. All kubernetes resources have a "resourceVersion" field as part of their metadata. This resourceVersion is a string that identifies the internal version of an object that can be used by clients to determine when objects have changed. When a record is about to be updated, it's version is checked against a pr ...
```
Changes to api-conventions.md LGTM. Thanks.

bgrant0607 commented on pkg/api/v1beta2/types.go:
```diff
@@ -861,7 +861,7 @@ type ObjectReference struct {
 	UID             types.UID `json:"uid,omitempty" description:"uid of the referent"`
 	APIVersion      string    `json:"apiVersion,omitempty" description:"API version of the referent"`
-	ResourceVersion string    `json:"resourceVersion,omitempty" description:"specific resourceVersion to which this reference is made, if any"`
+	ResourceVersion string    `json:"resourceVersion,omitempty" description:"string that identifies the internal version of this object that can be used by clients to determine when objects have changed; value must be treated as opaque by clients and passed unmodified back to the server: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conventions.md#concurrency-control-and-con ...
```
The only change here should be to add the URL to the doc, correct?

yujuhong commented:
@bgrant0607, I believe it was me. Yes, client isn't allowed to change the resourceVersion. However, in this case, even though the client got the ObjectReference previously from the server, they still have to *provide* an ObjectReference (including a resourceVersion) in this request. For example, the apiserver is not going to set the ObjectReference for you when you send an event. IMO, marking this read-only is a little bit confusing since it's different from all other cases, where clients should not set the field at all (e.g. creationTimestamp).

bgrant0607 commented:
@yujuhong That's true for virtually all fields that cannot be updated. PUT => replace.

bgrant0607 commented:
In case it wasn't clear, resourceVersion IS read-only, and should be documented as such. We can further clarify the overview of how updates work in api-conventions.md, if you feel it's necessary.

yujuhong commented:
@bgrant0607, the difference is whether you are creating or updating an object. In this case, when creating an event, you must provide an ObjectReference (with resourceVersion, etc). When you are updating, you don't necessarily have to. All other read-only fields do *not* need to be provided at the time of creation as they are populated by the system (in fact, it's better to leave the fields empty). As for updates, It might even make sense sometimes to update ObjectReference to point to another object.

That said, I agree that it can be argued either way, and I am fine with marking resourceVersion read-only.

bprashanth commented:
Since we're just proxying a read only value in the reference and I don't want to encourage hand setting the value (they should be using [GetReference] (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/ref.go#L43) anyway) I'm going with read-only.

lavalamp commented:
This LGTM as is.

The ResourceVersion in ObjectReference is just an ordinary field, it doesn't have the same semantics as the ResourceVersion in ObjectMeta.

bgrant0607 commented on pkg/api/v1beta2/types.go:
```diff
@@ -879,7 +879,7 @@ type ObjectReference struct {
 	UID             types.UID `json:"uid,omitempty" description:"uid of the referent"`
 	APIVersion      string    `json:"apiVersion,omitempty" description:"API version of the referent"`
-	ResourceVersion string    `json:"resourceVersion,omitempty" description:"specific resourceVersion to which this reference is made, if any"`
+	ResourceVersion string    `json:"resourceVersion,omitempty" description:"string that identifies the internal version of this object that can be used by clients to determine when objects have changed; populated by the system, read-only; value must be treated as opaque by clients and passed unmodified back to the server: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conven ...
```
As @lavalamp pointed out, this field isn't read-only.

bgrant0607 replied:
Please revert to the original text, actually.

bgrant0607 commented on pkg/api/v1beta3/types.go:
```diff
@@ -1111,7 +1111,7 @@ type ObjectReference struct {
 	UID             types.UID `json:"uid,omitempty"`
 	APIVersion      string    `json:"apiVersion,omitempty"`
-	ResourceVersion string    `json:"resourceVersion,omitempty"`
+	ResourceVersion string    `json:"resourceVersion,omitempty" description:"string that identifies the internal version of this object that can be used by clients to determine when objects have changed; populated by the system, read-only; value must be treated as opaque by clients and passed unmodified back to the server: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conven ...
```
Use the original text from v1beta1/2. Adding the link to the doc is ok.

bprashanth commented:
I'm Ok with a white lie if it'll stop people from shooting themselves in the foot. But meh, PTAL :)

bgrant0607 commented:
LGTM
